### ISSUE TITLE
Infinity loops inside akka

### ISSUE THREAD

**Issue by catap** [akka/akka#28601]
Hey,

I do have a akka-cluster that is working well for last couple of years and for last few days is started consume a lot of CPU, and as result it is defently blocked.

The code related to akka is rare changed (last time in December, before in June), and the system just works until the hell is going on.

It has two symptoms:
1) Almost each node has infinity loop with stack trace
```
"System-scheduler-1" #42 prio=5 os_prio=0 tid=0x00007f900dd820d0 nid=0x3d runnable [0x00007f8f6fffe000]
   java.lang.Thread.State: RUNNABLE
        at akka.dispatch.AbstractNodeQueue.pollNode(AbstractNodeQueue.java:163)
        at akka.actor.LightArrayRevolverScheduler$$anon$3.executeBucket$1(LightArrayRevolverScheduler.scala:277)
        at akka.actor.LightArrayRevolverScheduler$$anon$3.nextTick(LightArrayRevolverScheduler.scala:289)
        at akka.actor.LightArrayRevolverScheduler$$anon$3.run(LightArrayRevolverScheduler.scala:241)
        at java.lang.Thread.run(Thread.java:748)
```

2) Sometimes it has second thread that also looping at same code:
```
"System-akka.io.pinned-dispatcher-22" #72 prio=5 os_prio=0 tid=0x00007fbb4c0dde70 nid=0x57 runnable [0x00007fbba4ae9000]
   java.lang.Thread.State: RUNNABLE
        at akka.dispatch.AbstractNodeQueue.pollNode(AbstractNodeQueue.java:163)
        at akka.dispatch.AbstractNodeQueue.poll(AbstractNodeQueue.java:138)
        at akka.util.SerializedSuspendableExecutionContext.run$1(SerializedSuspendableExecutionContext.scala:69)
        at akka.util.SerializedSuspendableExecutionContext.run(SerializedSuspendableExecutionContext.scala:77)
        at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:41)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
```

As blind shot I've tried to upgrade to last akka 2.5.26 from 2.5.14 but it doesn't help.

It runs on 
```
openjdk version "1.8.0_242"
OpenJDK Runtime Environment (build 1.8.0_242-b08)
OpenJDK 64-Bit Server VM (build 25.242-b08, mixed mode)
```

from fedora 31 and uses Shenandoah as GC.

Technically a cause for this issue might be related to http://hg.openjdk.java.net/jdk/jdk/rev/0608ef3a7740 but it is back ported long time ago.

**Comment by catap** [comment]
After #28603 it still has some strange loops that consume a lot of CPU
```
"System-akka.io.pinned-dispatcher-31" #74 prio=5 os_prio=0 tid=0x00007f98fc02cf00 nid=0x5c runnable [0x00007f98bc005000]
   java.lang.Thread.State: RUNNABLE
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1163)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)

"System-akka.io.pinned-dispatcher-32" #73 prio=5 os_prio=0 tid=0x00007f99380232d0 nid=0x5b runnable [0x00007f98bc106000]
   java.lang.Thread.State: RUNNABLE
        at java.util.concurrent.LinkedBlockingQueue.signalNotEmpty(LinkedBlockingQueue.java:174)
        at java.util.concurrent.LinkedBlockingQueue.offer(LinkedBlockingQueue.java:430)
        at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1371)
        at akka.dispatch.ExecutorServiceDelegate.execute(ThreadPoolBuilder.scala:217)
        at akka.dispatch.ExecutorServiceDelegate.execute$(ThreadPoolBuilder.scala:217)
        at akka.dispatch.Dispatcher$LazyExecutorServiceDelegate.execute(Dispatcher.scala:42)
        at akka.dispatch.Dispatcher.executeTask(Dispatcher.scala:80)
        at akka.dispatch.MessageDispatcher.unbatchedExecute(AbstractDispatcher.scala:154)
        at akka.dispatch.BatchingExecutor.execute(BatchingExecutor.scala:125)
        at akka.dispatch.BatchingExecutor.execute$(BatchingExecutor.scala:116)
        at akka.dispatch.MessageDispatcher.execute(AbstractDispatcher.scala:88)
        at akka.util.SerializedSuspendableExecutionContext.attach(SerializedSuspendableExecutionContext.scala:81)
        at akka.util.SerializedSuspendableExecutionContext.remState(SerializedSuspendableExecutionContext.scala:50)
        at akka.util.SerializedSuspendableExecutionContext.run(SerializedSuspendableExecutionContext.scala:78)
        at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:41)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
```

**Comment by johanandren** [comment]
Since you are seeing the issue still, do you have a repeater for it or you just got (un)lucky that it happened again? 
Is that still on 2.5.x?
Any specific configuration of the dispatchers (very few threads?)

**Comment by catap** [comment]
@johanandren it is 2.5.26

Yes, I can easy reproduce it at one of clusters with do nothing. Just deploy, wait some time and the hell is going on.

The same code runs at dozen clusters but issue at only one. The cluster is split over two sub-DC at the provider as 50/50, and probably it has some network latency between it.

It hasn't got any specific configuration. The cluster's config:
```
akka {
  loggers = [de.heikoseeberger.akkalog4j.Log4jLogger]
  logging-filter = de.heikoseeberger.akkalog4j.Log4jLoggingFilter
  loglevel = "INFO"
  log-dead-letters-during-shutdown = false

  actor {
    provider = cluster
    warn-about-java-serializer-usage = false
  }

  cluster {
    roles = [bla-bla]
    auto-down-unreachable-after = 300s
  }

  remote {
    netty.tcp = {
      hostname = ${EXTERNAL_ADDR}
      bind-hostname = "0.0.0.0"
    }
  }

  management {
    cluster.bootstrap {
      form-new-cluster = on
      contact-point-discovery {
        service-name = ${AKKA_DISCOVERY_SERVICE_NAME}
        required-contact-point-nr = 1
      }
      join-decider {
        class = "GreedyLowestAddressJoinDecider"
      }
    }
    http {
      hostname = ${EXTERNAL_ADDR}
      bind-hostname = "0.0.0.0"
    }
  }

  discovery {
    method = akka-dns
  }
}
```

`GreedyLowestAddressJoinDecider` returns all nodes from and not only 5 as it hardcoded at `LowestAddressJoinDecider`.

Anyway, since #28603 CPU usage signifinlty decreased but without any loads one-two threads is consuming a lot of CPU and I've reported it here: https://github.com/akka/akka/issues/28601#issuecomment-585752246

**Comment by johanandren** [comment]
How are you measuring what is taking CPU in those cases?

**Comment by catap** [comment]
@johanandren over `top -H -p [jvm pid]`

After that I can see threads (task in linux kernel wording) and its ID and recovery stacktrace of each of them from `jstack` because `nid` is the same with value at top (but in HEX format).

**Comment by johanandren** [comment]
Would be great if you could use something like async-profiler (https://github.com/jvm-profiling-tools/async-profiler) to verify, because that hot-spin in `AbstractNodeQueue` you have modified in the PR should only ever be entered with a concurrent write and stop spinning as soon as the writer makes progress, so it seems very strange that it would dominate anything.

**Comment by catap** [comment]
@johanandren I don't think that using async-profiler is possibly here (because it runs over docker), but I will think about some proof.

Meanwhile, I'm upgrading this cluster to `2.5.27` and akka-managment `1.0.5`

**Comment by patriknw** [comment]
The `System-scheduler-1` thread is supposed to always be active. It's the timer scheduler.

It's not hot-spinning though. It should mostly sleep https://github.com/akka/akka/blob/master/akka-actor/src/main/scala/akka/actor/LightArrayRevolverScheduler.scala#L87

If your stack dumps are not mostly showing that it sleeps but shows that it is in that `executeBucket` method it might mean that there are very many scheduled tasks.

`akka.io.pinned-dispatcher` is the is the selector for the Tcp actor. Do you use Artery TCP? From the configuration it doesn't look like that, still netty.tcp. What else could be using TCP IO? Any TcpStream in your application?

Can you figure out what it is doing by turning on DEBUG level logging?

**Comment by catap** [comment]
@patriknw you are defently right, it is netty.tcp.

This application use akka only as a way to communicate between nodes in the cluster. It sends a bit messages, let say few hundreds per seconds. The application logic is outside of akka.

Mainly it uses call a function to broadcast a message to another node (by role), and it has actor that accepts this message and puts into queue to process by application thread.

No akka-http and another things. Just akka, akka-managment and it's all.

This application has jprofile as a way to profile and debug it, and I've enabled it at one of nodes to get this screenshot as proof that we have issue where I've explained.

<img width="1324" alt="Screenshot 2020-02-13 20 18 26" src="https://user-images.githubusercontent.com/37775/74470060-19086000-4e9e-11ea-8486-8105a1be63fe.png">

Cc @johanandren

**Comment by catap** [comment]
And one more things.

I've added a way to shutdown the during process life. It is shutdowning as:

```scala
    Await.result(akkaManagement.start(), Duration.Inf)
    Await.result(system.terminate(), Duration.Inf)
```

I've expected that as soon as both await is finished, all akka-related thread is dissapears.

Anyway, it is still exists, and it still running.

You can see cutted output of jstack here: https://gist.github.com/catap/f2d46e1fde449509a5fb410240d3fb81

**Comment by patriknw** [comment]
Thanks for investigating. May I suggest that we take this systematically step-by-step. We are not sure that the suggestions in the PR are the right thing to do, and before we even start reviewing that in depth we would like to understand the root problem.

Could you run without the changes in the PR and instead add some additional logging or println that shows what is going on.

**Comment by catap** [comment]
@patriknw unfortunately enable debug log isn't easy task to be honest.

**Comment by patriknw** [comment]
`SerializedSuspendableExecutionContext` is only used by Akka IO. That is used by things like Akka HTTP, Streams TCP, Artery TCP. That is used by many many users and I'm just very surprised that we would not haven't heard of this problem before. That is why I'm trying to understand what your application is doing to trigger this problem.

Akka Management is using Akka HTTP. So, is that the only usage of Akka IO in your application? Does the problem disappear if you remove Akka Management?

Can you reproduce it in a smaller application that we can run?

**Comment by catap** [comment]
@patriknw I've tried.

Right now I've discovered that easy way to reproduce it is restart akka system.

So, right now the case to reproduce it something like that:
1. Deploy cluster and allow it to join it
2. restart akka system
3. after rejoin some nodes will have CPU leak. Without my hack leak much worst, with it exist but not so crazy.

If I'm split (3) for two step: shutdown akka and wait start and spend some time, chances to have this loop is smaller on machine but at cluster with issue it still exists.

Also I saw in log a error related to "pool timeout" because it was longer than `akka.http.host-connection-pool.idle-timeout`, but I think it is related to https://github.com/akka/akka-http/issues/2728 and just warning.

**Comment by catap** [comment]
I can reproduce it locally at one instance. Without any hacks inside akka. So, it is vanilla.

I've added `seed-nodes` with correct IP/PORT but with wrong akka system name. For example as soon as it started it it has infinity loop.

The akka in debug log:
```
2020-02-15 19:26:00 DEBUG EventStream: - logger log1-Log4jLogger started
2020-02-15 19:26:00 DEBUG EventStream: - Default Loggers started
2020-02-15 19:26:00 INFO  Remoting: - Starting remoting
2020-02-15 19:26:01 INFO  Remoting: - Remoting started; listening on addresses :[akka.tcp://System@10.254.254.201:2552]
2020-02-15 19:26:01 INFO  Remoting: - Remoting now listens on addresses: [akka.tcp://System@10.254.254.201:2552]
2020-02-15 19:26:01 INFO  Cluster(akka://System): - Cluster Node [akka.tcp://System@10.254.254.201:2552] - Starting up, Akka version [2.5.27] ...
2020-02-15 19:26:01 INFO  Cluster(akka://System): - Cluster Node [akka.tcp://System@10.254.254.201:2552] - Registered cluster JMX MBean [akka:type=Cluster]
2020-02-15 19:26:01 INFO  Cluster(akka://System): - Cluster Node [akka.tcp://System@10.254.254.201:2552] - Started up successfully
2020-02-15 19:26:01 WARN  AutoDown: - Don't use auto-down feature of Akka Cluster in production. See 'Auto-downing (DO NOT USE)' section of Akka Cluster documentation.
2020-02-15 19:26:01 DEBUG EndpointWriter: - Associated [akka.tcp://System@10.254.254.201:2552] -> [akka.tcp://NonExistSystem@10.254.254.201:2552]
2020-02-15 19:26:01 DEBUG Remoting: - Associated [akka.tcp://System@10.254.254.201:2552] <- [akka.tcp://System@10.254.254.201:2552]
2020-02-15 19:26:02 DEBUG Serialization(akka://System): - Using serializer [akka.cluster.protobuf.ClusterMessageSerializer] for message [akka.cluster.InternalClusterAction$InitJoin]
2020-02-15 19:26:02 INFO  ClusterBootstrap: - Old `form-new-cluster` property set. Using value {} as `new-cluster-enabled` and ignoring `new-cluster-enabled`. Please update to using `new-cluster-enabled`
2020-02-15 19:26:02 DEBUG DnsServiceDiscovery: - system resolver is not async-dns. Loading isolated resolver
2020-02-15 19:26:02 INFO  ActorSystemImpl: - Creating async dns resolver async-dns with manager name SD-DNS
2020-02-15 19:26:02 DEBUG EndpointWriter: - Drained buffer with maxWriteCount: 50, fullBackoffCount: 1, smallBackoffCount: 0, noBackoffCount: 0 , adaptiveBackoff: 1000
2020-02-15 19:26:02 INFO  ClusterBootstrap: - Bootstrap using default `akka.discovery` method: DnsServiceDiscovery
2020-02-15 19:26:02 ERROR EndpointWriter: - dropping message [class akka.actor.ActorSelectionMessage] for non-local recipient [Actor[akka.tcp://NonExistSystem@10.254.254.201:2552/]] arriving at [akka.tcp://NonExistSystem@10.254.254.201:2552] inbound addresses are [akka.tcp://System@10.254.254.201:2552]
2020-02-15 19:26:02 INFO  HealthChecksImpl: - Loading readiness checks List(NamedHealthCheck(cluster-membership,akka.management.cluster.scaladsl.ClusterMembershipCheck))
2020-02-15 19:26:02 INFO  HealthChecksImpl: - Loading liveness checks List()
2020-02-15 19:26:02 DEBUG AsyncDnsResolver: - Using name servers [List(/127.0.0.11:53)] and search domains [List()] with ndots=0
2020-02-15 19:26:02 DEBUG TimerScheduler: - Start timer [CacheCleanup] with generation [1]
2020-02-15 19:26:02 DEBUG UdpListener: - Successfully bound to [/0.0.0.0:36148]
2020-02-15 19:26:02 DEBUG DnsClient: - Bound to UDP address [/0.0.0.0:36148]
2020-02-15 19:26:02 INFO  AkkaManagement: - Binding Akka Management (HTTP) endpoint to: 0.0.0.0:8558
2020-02-15 19:26:02 INFO  AkkaManagement: - Including HTTP management routes for ClusterHttpManagementRouteProvider
2020-02-15 19:26:02 INFO  AkkaManagement: - Including HTTP management routes for ClusterBootstrap
2020-02-15 19:26:02 INFO  ClusterBootstrap: - Using self contact point address: http://10.254.254.201:8558
2020-02-15 19:26:03 INFO  AkkaManagement: - Including HTTP management routes for HealthCheckRoutes
2020-02-15 19:26:04 WARN  ClusterBootstrap: - Application is configured with specific `akka.cluster.seed-nodes`: [akka.tcp://NonExistSystem@10.254.254.201:2552], bailing out of the bootstrap process! If you want to use the automatic bootstrap mechanism, make sure to NOT set explicit seed nodes in the configuration. This node will attempt to join the configured seed nodes.
2020-02-15 19:26:04 DEBUG TcpListener: - Successfully bound to /0.0.0.0:8558
2020-02-15 19:26:04 INFO  AkkaManagement: - Bound Akka Management (HTTP) endpoint to: 0.0.0.0:8558
```

At jprofile it looks like:

<img width="809" alt="Screenshot 2020-02-15 20 27 54" src="https://user-images.githubusercontent.com/37775/74593999-acbe6580-5031-11ea-9cbb-6a62e049d452.png">


I believe that it is something close to my case because my application can restart akka, and if remote system is in GC loop, this case (when remote system doesn't available) may be a case.

**Comment by catap** [comment]
and the same values from debugger inside IDEA.

<img width="1604" alt="Screenshot 2020-02-16 21 30 13" src="https://user-images.githubusercontent.com/37775/74612233-a7315000-5103-11ea-9a48-024916323874.png">

**Comment by catap** [comment]
And I've added a bit logging into `AbstractNodeQueue` that you can see here https://gist.github.com/catap/286be063162dc6ce4f711641f8c5d671

The loop happened at
```
2020-02-16 22:20:09 DEBUG AbstractNodeQueue: - (akka.dispatch.AbstractNodeQueue$Node@51a9c0a9) pollNode, next: null; tail: akka.dispatch.AbstractNodeQueue$Node@57ad5ef4; get(): akka.dispatch.AbstractNodeQueue$Node@51a9c0a9
```

Why? I see same `tail` and `this` at debugger and `next` is `null` as it should be.

**Comment by patriknw** [comment]
Now you are back to the `LightArrayRevolverScheduler` again. Previously it was `SerializedSuspendableExecutionContext`.

```
Created: akka.actor.LightArrayRevolverScheduler$TaskQueue
```

That is expected. It's not a hot spinning loop. It's sleeping during the `tick-duration`, which by default is 10 ms. You can confirm by adding a println before waitNanos in LightArrayRevolverScheduler.

This should normally not add much load to the system, but if you want to reduce it you can change the `tick-duration` to 30 ms. Don't increase above that because then the Cluster will anyway create a different scheduler with 33ms tick-duration, which is needed for failure detection.
```
akka.scheduler.tick-duration = 30ms
```

**Comment by catap** [comment]
@patriknw first at all the first stacktracke contains `LightArrayRevolverScheduler`, so, it is still the same bug.

At my simple case it looping for infinity time and thread with this loop is consuming almost all CPU because the system does nothing, and just started up.

You can see a video with `top -H -p [pid]` that confirms that I see: https://www.dropbox.com/s/6qt7o65hfb1vybb/wrong%20system.mov?dl=0

If I change system at seed nodes to something that exist it hasn't got this loop: https://www.dropbox.com/s/9swv7p7itpk128c/correct%20system.mov?dl=0

So, if you have an idea how can I proof for you that the bug exists, I'm ready do it.

**Comment by patriknw** [comment]
That looks convincing. I tried to reproduce by joining with wrong system name. I see same logs regarding "non-local recipient" as expected, but it never spins in that `do-while` loop in `pollNode` (or `peekNode`).

I can simulate triggering of the spinning by adding a sleep in `add`, between the two operations `getAndSet(n).setNext(n)`. The consumer in `pollNode` will progress when the producer (`add`) has completed the operation. In other words, I don't see the infinite spinning in `pollNode`.

The mystery is still why the producer thread can't progress and release the consumer in your case? Have you tried with other JDK (and other GC than Shenandoah)?

**Comment by catap** [comment]
@patriknw this is very fragile reproduction bug to be honest.

I can reproduced it only inside this application after it start and build index from DB. Base on nature of application DB is changing continuously and index size, and number of object is very variety, I can safley assume that it changes ±50% from average size.

After it build it should warmup its caches and do some thing like that.

After that it creates an akka system and tries join the cluster. Here I can reproduce it locally.

I've tried to move to another GC, but it cover the bug. Honestly any self-verification option of Shenandoah is cover this bug also.

I can reproduce it on:
 - stock openjdk 8 from fedora 31
 - release and fastdebug build from http://hg.openjdk.java.net/shenandoah/jdk8

But as soon as I switch off any concurrency or slow everything down by self-verification my case is falling apart.

If I change GC that makes bigger pauses, it also doesn't reproduced.

I'm really afraid that it is something like wired side effect of race condition inside of akka, because you're using sleeps over `Thread.sleep` for small time, and it isn't so precise. Honestly, at JVM is very difficult to achieve a predicted sleep, you can read a lot of details about it [here](https://medium.com/@catap/sleeping-at-jvm-thread-for-a-very-short-period-107339eaf8d5) and [here](https://medium.com/@catap/sleeping-at-jvm-thread-for-a-very-short-period-at-multi-threads-environment-bcbf457e929e) (disclamer: I've wrote two of this aricles).

Otherwise I see your point:

> I can simulate triggering of the spinning by adding a sleep in add, between the two operations getAndSet(n).setNext(n). 

It is good point to be honest. Nobody can guarantee when OS switch threads and on havy loaded system or with a lot of threads, thread definitely might be suspended between `getAndSet(..)` and `setNext(..)`.

In my case Shenandoah has background thread that clean everything up and probably cause this sleep.

**Comment by patriknw** [comment]
Do I read that correctly that you can only reproduce it with Shenandoah enabled?

> But as soon as I switch off any concurrency or slow everything down by self-verification my case is falling apart.

What exactly do you mean by "switch off any concurrency" and "self-verification"?

Do you mean that if you add a `yield` or `sleep` inside the `do-while` loop the problem disappears? And then I mean that the consumer exits the loop, and not only loops more slowly.

**Comment by catap** [comment]
> Do I read that correctly that you can only reproduce it with Shenandoah enabled?

I guess so.

I can't reproduce it if I:
 - switch off Shenandoah and move to G1 that adds bigger pauses in fact
 - use slow-debug build of openjdk
 - adds options that slowdown Shenandoah such as: `-XX:+ShenandoahVerify` or `-XX:+ShenandoahVerifyOptoBarriersq` or `XX:VerifyStrictOopOperations`
 - or disable part of JVM compiler over `-Xint` or `-XX:TieredStopAtLevel=1` or `-XX:-TieredCompilation`.

I feel that you tried to say that it is a Shenandoah related bug. Perhaps. But all another (non Shenandoah) GC makes everything slower, and when I slow everything down by Shenandoah this also help to remove symptoms. I really feel that it might be some sort of race condition or something like that.

And yes, if I add `Thread.yield()` or `Thread.sleep(0)`  inside `do-while` at `pollNode` it definitely helps to remove this symptoms because this loop won't consume CPU anymore, because it will have a sleep inside.

**Comment by patriknw** [comment]
> And yes, if I add Thread.yield() or Thread.sleep(0) inside do-while at pollNode it definitely helps to remove this symptoms because this loop won't consume CPU anymore, because it will have a sleep inside.

Can you verify somehow that it actually exits the loop. Just slowing down the loop but anyway looping forever is not a solution. Maybe you can verify that by adding a counter and print something every 10000 lap of the loop?

**Comment by catap** [comment]
@patriknw very good idea, thanks.

Adding something `if (counter % 1000 == 0) logger.debug...` broke my case — as I said it is very fragile.

Anyway, adding just a counter and log that reports how many spins it did isn't.

So, I'd like to share with you:
1) debug log with reported counters how many spins this loop made: https://gist.github.com/catap/9ac6699becf55bb10d4e05362ce28e27
2) a picture from attached debuger that shown a current value of this counte plus a code that is used to report it:

<img width="839" alt="Screenshot 2020-02-17 14 18 33" src="https://user-images.githubusercontent.com/37775/74657526-b3afba00-5190-11ea-8b35-e81fc40648e7.png">

I guess it proofs that it is looping forever.

**Comment by patriknw** [comment]
ok, then I suggest that we add a yield if counter is above some value.
```
          long counter = 0;
          do {
              next = tail.next();
              // When dropping support for JDK 8 we can add
              // Thread.onSpinWait();
              counter += 1;
              if (counter > 10000L)
                  Thread.yield();
          } while (next == null);
```

**Comment by catap** [comment]
@patriknw sure, but at this case it won't fix an issue at all:

<img width="808" alt="Screenshot 2020-02-17 14 47 05" src="https://user-images.githubusercontent.com/37775/74659331-69303c80-5194-11ea-8d2f-6451d56dcfae.png">

I can add `Thread.sleep()` that adds bigger slepe, but it still cover an issue.

For my point of view this loop is a symptoms for real issue.

**Comment by patriknw** [comment]
I see it as a workaround for unfair thread scheduling by the OS or JVM in your environment.

**Comment by catap** [comment]
@patriknw but this loop will spin forever, isn't it?

Anyway, it seems very similar to https://github.com/akka/akka/pull/28603 and as I pointed before, I do have looping at another points: https://github.com/akka/akka/issues/28601#issuecomment-585752246

But I don't know how to reproduce it on smaller environment: not at the cluster with real customer for few dozens of machines.

**Comment by patriknw** [comment]
> but this loop will spin forever, isn't it?

If I understood you correctly the problem disappears when you add such delay (or println) in the loop, i.e. it gives the producer thread a chance to run and complete its operation and thereby the consumer loop will exit.

**Comment by catap** [comment]
@patriknw the issue is disappear from my test case.

The cluster is runs with https://github.com/akka/akka/pull/28603 that adds similar delay.

It "solves" this loop, but some nodes still has another loops: https://github.com/akka/akka/issues/28601#issuecomment-585752246 but this one appears not so often, and it is much more difficult to achieve on the cluster.

Anyway, I still have a sort of feeling that all theses issue is related to something common, and better to hunt it down, and not just cover it by sleep.

**Comment by patriknw** [comment]
I agree that we want to fix the bug if it's a bug. I think everything from our recent discussion pointed to unfair thread scheduling, but ok if it doesn't solve it then then it's not a solution. I'm out of ideas for now.

**Comment by catap** [comment]
@patriknw JFYI: I've reported it into shenandoah-dev@openjdk.java.net because I definitely need their advice to eliminate or confirm that it is Shenandoah bug.

**Comment by catap** [comment]
@patriknw just an idea I've tried to stop thread with infinity loop by adding `Thread.stop()` after `system.terminate()` is complete:

```scala
        cluster.leave(cluster.selfAddress)
        system.terminate() onComplete { _ =>
          val threadName = system.name + "-"
          Thread.getAllStackTraces().keySet().asScala
            .filter(_.getName.startsWith(threadName))
            .foreach { thread =>
              try {
                thread.stop()
              } catch {
                case e: Throwable =>
                // do nothing
              }
            }
          promise.success(Done)
        }
        promise.future
```

As result akka crashed (expected) but I've expected that after it is terminated, it has less threads: https://gist.github.com/catap/f8c30243a5f554fcfff8c5d8fa86d906

**Comment by catap** [comment]
@patriknw status update here.

I can proof that it is GC related bug. As soon as I simplified it to https://gist.github.com/catap/f753e5d12409be8ab19135b40769c3db it creates a log like:
```
DEBUG AbstractNodeQueue: - [setNext] (Node(2576)) ; newNext: Node(3179); next: Node(3179); next(): Node(3179)
DEBUG AbstractNodeQueue: - [pollNode] (Node(2576)); tail.next(): null; tail.next: null
```

that can't happened without GC bug.

At my application I've reorganize a bit index to reduce requirements to heap, as result it decreased number of GC loop and decreased probability of this loop.

**Comment by patriknw** [comment]
Thanks for investigating. Does it mean that we can close this issue?

**Comment by catap** [comment]
@patriknw right now the code deployed with `yield` hack and one more that brake the loop and return null if number of iterations is huge.

Anyway, it creates loop inside `SerializedSuspendableExecutionContext` that also creates some pain. Not so big as `LightArrayRevolverScheduler` but...

I'm waiting until GC bug is fixed, and after that I'll confirm that it is disappear, or I will continue digging that's going on.

**Comment by catap** [comment]
It looks like that the cause of this issue was fixed inside used GC: https://mail.openjdk.java.net/pipermail/shenandoah-dev/2020-February/011516.html

I close this issue and pull request because it is definitely do not related to akka.

**Comment by johanandren** [comment]
Thanks for letting us know and good that the root issue was fixed!
